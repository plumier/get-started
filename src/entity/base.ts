import { authorize, entity } from "plumier"
import { Column, CreateDateColumn, Entity, PrimaryGeneratedColumn, UpdateDateColumn } from "typeorm"

@Entity()
export class BaseEntity {
    @authorize.readonly()
    @PrimaryGeneratedColumn()
    id: number

    @authorize.readonly()
    @CreateDateColumn()
    createdAt: Date

    @authorize.readonly()
    @UpdateDateColumn()
    updatedAt: Date

    @entity.deleteColumn()
    @Column({ default: false })
    deleted: boolean
}

export interface LoginUser {
    userId: number,
    role: "User" | "Admin"
}
