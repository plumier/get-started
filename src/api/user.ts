import bcrypt from "bcryptjs"
import { authorize, entityPolicy, preSave, Public, route, val } from "plumier"
import { Column, Entity } from "typeorm"

import { AppRole, BaseEntity } from "./_shared"

@route.controller(c => {
    c.post().authorize(Public)
    c.actions("Put", "Patch", "Delete").authorize(AppRole.Owner, AppRole.Admin)
})
@Entity()
export class User extends BaseEntity {
    @val.required()
    @val.unique()
    @val.email()
    @authorize.read(AppRole.Owner, AppRole.Admin)
    @Column()
    email: string

    @val.required()
    @authorize.writeonly()
    @Column()
    password: string

    @val.required()
    @Column()
    name: string

    @authorize.write(AppRole.Admin)
    @Column({ default: AppRole.User })
    role: AppRole

    @preSave()
    async hashPassword() {
        if (this.password)
            this.password = await bcrypt.hash(this.password, await bcrypt.genSalt())
    }
}

export const UserOwnerPolicy = entityPolicy(User)
    .define(AppRole.Owner, (ctx, id) => ctx.user?.userId === id)
