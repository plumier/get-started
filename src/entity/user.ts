import bcrypt from "bcryptjs"
import { authorize, preSave, Public, route, val } from "plumier"
import { Column, Entity } from "typeorm"

import { BaseEntity } from "./base"

@route.controller(c => {
    c.post().authorize(Public)
    c.actions("Put", "Patch", "Delete").authorize("Owner", "Admin")
})
@Entity()
export class User extends BaseEntity {
    @val.required()
    @val.unique()
    @val.email()
    @authorize.read("Owner", "Admin")
    @Column()
    email: string

    @val.required()
    @authorize.writeonly()
    @Column()
    password: string

    @val.required()
    @Column()
    name: string

    @authorize.write("Admin")
    @Column({ default: "User" })
    role: "User" | "Admin"

    @preSave()
    async hashPassword() {
        if (this.password)
            this.password = await bcrypt.hash(this.password, await bcrypt.genSalt())
    }
}
