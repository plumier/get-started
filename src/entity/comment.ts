import { authorize, bind, preSave, val } from "plumier"
import { Column, Entity, ManyToOne } from "typeorm"

import { BaseEntity, LoginUser } from "./base"
import { Todo } from "./todo"
import { User } from "./user"

@Entity()
export class Comment extends BaseEntity {
    @val.required()
    @Column()
    message: string

    @ManyToOne(x => User)
    user: User

    @authorize.readonly()
    @ManyToOne(x => Todo)
    todo: Todo

    @preSave("post")
    setUser(@bind.user() user: LoginUser) {
        // set current login user as the user of the comment
        this.user = <User>{ id: user.userId }
    }
}
