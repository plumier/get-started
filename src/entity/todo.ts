import { authorize, bind, preSave, route, val } from "plumier"
import { Column, Entity, ManyToOne, OneToMany } from "typeorm"

import { BaseEntity, LoginUser } from "./base"
import { Comment } from "./comment"
import { User } from "./user"

@route.controller(c => c.actions("Put", "Patch", "Delete").authorize("Owner", "Admin"))
@Entity()
export class Todo extends BaseEntity {
    @val.required()
    @Column()
    message: string

    @authorize.write("Owner", "Admin")
    @Column({ default: false })
    completed: boolean

    @authorize.readonly()
    @ManyToOne(x => User)
    user: User

    @route.controller(c => c.actions("Put", "Patch", "Delete").authorize("Owner", "Admin"))
    @OneToMany(x => Comment, x => x.todo)
    comments: Comment[]

    @preSave("post")
    setUser(@bind.user() user: LoginUser) {
        // set current login user as the user of the todo
        this.user = <User>{ id: user.userId }
    }
}
