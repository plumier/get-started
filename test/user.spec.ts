import "@plumier/typeorm"

import supertest from "supertest"
import { getConnection, getManager } from "typeorm"

import { User } from "../src/entity/user"
import createApp from "../src/app"
import dotenv from "dotenv"
import { join } from "path"
import bcrypt from "bcryptjs"
import { adminToken, createToken, userToken, ignoreProps } from "./shared"
import { AppRole } from "../src/entity/base"

dotenv.config({ path: join(__dirname, ".env-test") })

afterEach(async () => {
    const con = getConnection()
    if (con.isConnected)
        await con.close()
})


describe("User Spec", () => {
    it("Should allow anyone to register", async () => {
        const app = await createApp({ mode: "production" })
        await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: "john0doe#",
                name: "John Doe",
            })
            .expect(200)
    })
    it("Should hash password during registration", async () => {
        const app = await createApp({ mode: "production" })
        const pwd = "john0doe#"
        const { body } = await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: pwd,
                name: "John Doe",
            })
            .expect(200)
        const repo = getManager().getRepository(User)
        const user = await repo.findOne(body.id)
        expect(await bcrypt.compare(pwd, user?.password!)).toBe(true)
    })
    it("Should not allow user to specify role", async () => {
        const app = await createApp({ mode: "production" })
        await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: "john0doe#",
                name: "John Doe",
                role: "Admin"
            })
            .expect(403)
    })
    it("Should not showing password on the response to anyone", async () => {
        const app = await createApp({ mode: "production" })
        const { body: user } = await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: "john0doe#",
                name: "John Doe",
            })
            .expect(200)
        const { body } = await supertest(app.callback())
            .get(`/api/v1/users/${user.id}`)
            .set("Authorization", `Bearer ${userToken}`)
            .expect(200)
        expect(body).toMatchSnapshot(ignoreProps)
    })
    it("Should only showing Email to the owner or admin on response", async () => {
        const app = await createApp({ mode: "production" })
        const { body: user } = await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: "john0doe#",
                name: "John Doe",
            })
            .expect(200)
        const token = createToken(user.id, AppRole.User)
        const { body: byItself } = await supertest(app.callback())
            .get(`/api/v1/users/${user.id}`)
            .set("Authorization", `Bearer ${token}`)
            .expect(200)
        const { body: bySomeoneElse } = await supertest(app.callback())
            .get(`/api/v1/users/${user.id}`)
            .set("Authorization", `Bearer ${userToken}`)
            .expect(200)
        const { body: byAdmin } = await supertest(app.callback())
            .get(`/api/v1/users/${user.id}`)
            .set("Authorization", `Bearer ${adminToken}`)
            .expect(200)
        expect(byItself).toMatchSnapshot(ignoreProps)
        expect(bySomeoneElse).toMatchSnapshot(ignoreProps)
        expect(byAdmin).toMatchSnapshot(ignoreProps)
    })
    it("Should not allow user to set their own role", async () => {
        const app = await createApp({ mode: "production" })
        await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: "john0doe#",
                name: "John Doe",
                role: "Admin"
            })
            .expect(403)
    })
    it("Should allow admin to set someone else role", async () => {
        const app = await createApp({ mode: "production" })
        const { body: user } = await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: "john0doe#",
                name: "John Doe",
            })
            .expect(200)
        await supertest(app.callback())
            .patch(`/api/v1/users/${user.id}`)
            .send({
                role: "Admin"
            })
            .set("Authorization", `Bearer ${adminToken}`)
            .expect(200)
    })
    it("Should not allow any user to modify other user", async () => {
        const app = await createApp({ mode: "production" })
        const { body: userOne } = await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: "john0doe#",
                name: "John Doe",
            })
            .expect(200)
        const { body: userTwo } = await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "jane.doe@gmail.com",
                password: "jane0doe#",
                name: "Jane Doe",
            })
            .expect(200)
        const userTwoToken = createToken(userTwo.id, AppRole.User)
        await supertest(app.callback())
            .patch(`/api/v1/users/${userOne.id}`)
            .send({
                role: "Admin"
            })
            .set("Authorization", `Bearer ${userTwoToken}`)
            .expect(401)
    })
    it("Should allow any user to delete their own account", async () => {
        const app = await createApp({ mode: "production" })
        const { body: userOne } = await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: "john0doe#",
                name: "John Doe",
            })
            .expect(200)
        const userOneToken = createToken(userOne.id, AppRole.User)
        await supertest(app.callback())
            .delete(`/api/v1/users/${userOne.id}`)
            .set("Authorization", `Bearer ${userOneToken}`)
            .expect(200)
        const { body } = await supertest(app.callback())
            .get(`/api/v1/users/${userOne.id}`)
            .set("Authorization", `Bearer ${userOneToken}`)
            .expect(200)
        expect(body.deleted).toBe(true)
    })
    it("Should allow admin to delete any account", async () => {
        const app = await createApp({ mode: "production" })
        const { body: userOne } = await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: "john0doe#",
                name: "John Doe",
            })
            .expect(200)
        await supertest(app.callback())
            .delete(`/api/v1/users/${userOne.id}`)
            .set("Authorization", `Bearer ${adminToken}`)
            .expect(200)
        const { body } = await supertest(app.callback())
            .get(`/api/v1/users/${userOne.id}`)
            .set("Authorization", `Bearer ${adminToken}`)
            .expect(200)
        expect(body.deleted).toBe(true)
    })
    it("Should not allow any user to delete someone else user", async () => {
        const app = await createApp({ mode: "production" })
        const { body: userOne } = await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "john.doe@gmail.com",
                password: "john0doe#",
                name: "John Doe",
            })
            .expect(200)
        const { body: userTwo } = await supertest(app.callback())
            .post("/api/v1/users")
            .send({
                email: "jane.doe@gmail.com",
                password: "jane0doe#",
                name: "Jane Doe",
            })
            .expect(200)
        const userTwoToken = createToken(userTwo.id, AppRole.User)
        await supertest(app.callback())
            .delete(`/api/v1/users/${userOne.id}`)
            .set("Authorization", `Bearer ${userTwoToken}`)
            .expect(401)
    })
})