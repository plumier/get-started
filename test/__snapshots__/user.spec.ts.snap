// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Spec Should not showing password on the response to anyone 1`] = `
Object {
  "createdAt": Any<String>,
  "deleted": false,
  "id": 1,
  "name": "John Doe",
  "role": "User",
  "updatedAt": Any<String>,
}
`;

exports[`User Spec Should only showing Email to the owner or admin on response 1`] = `
Object {
  "createdAt": Any<String>,
  "deleted": false,
  "email": "john.doe@gmail.com",
  "id": 1,
  "name": "John Doe",
  "role": "User",
  "updatedAt": Any<String>,
}
`;

exports[`User Spec Should only showing Email to the owner or admin on response 2`] = `
Object {
  "createdAt": Any<String>,
  "deleted": false,
  "id": 1,
  "name": "John Doe",
  "role": "User",
  "updatedAt": Any<String>,
}
`;

exports[`User Spec Should only showing Email to the owner or admin on response 3`] = `
Object {
  "createdAt": Any<String>,
  "deleted": false,
  "email": "john.doe@gmail.com",
  "id": 1,
  "name": "John Doe",
  "role": "User",
  "updatedAt": Any<String>,
}
`;
